test/validation/api/ipsec/ipsec.c

```diff
@@ -534,17 +536,36 @@ static int recv_pkts_inline(const ipsec_test_part *part,
 
 		ev = odp_queue_deq(suite_context.queue);
 		if (ODP_EVENT_INVALID != ev) {
+			odp_packet_t pkt;
+			int num_pkts = 0;
+
 			CU_ASSERT(odp_event_is_valid(ev) == 1);
-			CU_ASSERT_EQUAL(ODP_EVENT_PACKET,
-					odp_event_types(ev, &subtype));
-			CU_ASSERT_EQUAL(ODP_EVENT_PACKET_IPSEC,
-					subtype);
+			CU_ASSERT_EQUAL(ODP_EVENT_PACKET, odp_event_type(ev));
+			pkt = odp_packet_from_event(ev);
+
 			CU_ASSERT(!part->out[i].status.error.sa_lookup);
 
-			pkto[i] = odp_ipsec_packet_from_event(ev);
-			CU_ASSERT(odp_packet_subtype(pkto[i]) ==
-				  ODP_EVENT_PACKET_IPSEC);
-			i++;
+			reass_status = odp_packet_reass_status(pkt);
+			if (ODP_PACKET_REASS_INCOMPLETE != reass_status) {
+				pkto[i] = pkt;
+				num_pkts = 1;
+			} else {
+				odp_packet_t frags[MAX_FRAGS];
+				int j;
+
+				CU_ASSERT(0 ==
+					  odp_packet_reass_partial_state(pkt, frags, &reass_state));
+				num_pkts = reass_state.num_frags;
+
+				CU_ASSERT_FATAL(reass_state.num_frags <= i + MAX_FRAGS);
```

Comment:
The logic here is wrong. If i > 0 this would allow even more than MAX_FRAGS fragments.

A better assert would be this:

CU_ASSERT_FATAL(i + num_pkts <= part->num_pkt);

In addition to fixing the logic, it uses num_pkts instead of reass_state.num_frags because num_pkts is the loop bound in the loop that could overflow the array without this assert. Another improvement is to compare against part->num_pkt as that is the maximum number of packets that recv_pkts_inline() can return.

Reply:
Agreed. Will address in next version.